### PR TITLE
Clarity how to interpret GPURenderPassColorAttachment/clearValue

### PR DESCRIPTION
**Pull request by Jiawei-Shao**
fixes: https://github.com/gpuweb/gpuweb/issues/965

***
<a href="https://pr-preview.s3.amazonaws.com/Jiawei-Shao/gpuweb/pull/2795.html" title="Last updated on May 12, 2022, 6:02 AM UTC (377833f)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/gpuweb/gpuweb/2795/498a37e...Jiawei-Shao:377833f.html" title="Last updated on May 12, 2022, 6:02 AM UTC (377833f)">Diff</a>

### CODE REVIEW

File: spec/index.bs [hunk 1]
```diff
@@ -7960,6 +7960,27 @@ dictionary GPURenderPassColorAttachment {
         render pass. If not [=map/exist|provided=] defaults to `{r: 0, g: 0, b: 0, a: 0}`. Ignored
         if {{GPURenderPassColorAttachment/loadOp}} is not {{GPULoadOp/"clear"}}.
 
+        <div class=note>
```
Comment (Kangz) [inline spec/index.bs:7963]: Putting everything in a note means it is non-normative (just explanation that's not technically part of the standard). Did you mean to only put the first paragraph as a note?

**Review by Kangz — DISMISSED** [review]
LGTM but I'd like an editor to take a look as well.

File: spec/index.bs (hunk 1 above)
Comment (Jiawei-Shao) [reply spec/index.bs]: Oh I'd like to make these a part of SPEC and I have moved them out of "note".

File: spec/index.bs [hunk 2]
```diff
@@ -7960,6 +7960,25 @@ dictionary GPURenderPassColorAttachment {
...
+        The members of {{GPURenderPassColorAttachment/clearValue}} are all double values, so
+        they will first be converted to the fully qualified format type of
+        {{GPURenderPassColorAttachment/view}} before being set as the clear value of
+        {{GPURenderPassColorAttachment/view}}.
+
+        Let |colorAttachmentFormat| be
+        {{GPURenderPassColorAttachment/view}}.{{GPUTextureView/[[descriptor]]}}.{{GPUTextureViewDescriptor/format}}.
+        - If the components of |colorAttachmentFormat| are floating point values, the members of
+            {{GPURenderPassColorAttachment/clearValue}} will be converted into floating point
+            values.
```
Comment (kainino0x) [inline spec/index.bs:7972]: The clearValue is already floating point values (64-bit). The conversion is between different precisions of floating point values (e.g. f64->f16 or f64->rg11b10ufloat) and there's still ambiguity about a few things:
- is it truncated or rounded, and if so how
- what happens on underflow (f64 is smaller than the smallest target non-denormal value)
- what happens on overflow (f64 is larger than the largest target finite value)

The WebIDL conversions for f64->f32 for example specify this in detail:
https://webidl.spec.whatwg.org/#es-unrestricted-float

All that to say, we need to write our own conversion rules (similar to WebIDL's but generalized - as mentioned here https://github.com/gpuweb/gpuweb/issues/965#issuecomment-675072177) and we need to verify/test they actually work in the native APIs (results written aren't off-by-one).

File: spec/index.bs [hunk 3]
```diff
@@ -7960,6 +7960,25 @@ dictionary GPURenderPassColorAttachment {
...
+        - If the components of |colorAttachmentFormat| are signed or unsigned normalized values,
+            the members of {{GPURenderPassColorAttachment/clearValue}} will first be clamped
+            into the range [0, 1], then converted into floating point values, and finally
+            converted into the signed or unsigned normalized values.
```
Comment (kainino0x) [inline spec/index.bs:7976]: also needs to define rounding

File: spec/index.bs [hunk 4]
```diff
@@ -7960,6 +7960,25 @@ dictionary GPURenderPassColorAttachment {
...
+        - If the components of |colorAttachmentFormat| are signed or unsigned integer values,
+            the members of {{GPURenderPassColorAttachment/clearValue}} will first be clamped
+            into the range of signed or unsigned integer values with same bit length, and then
+            rounded into the nearest signed or unsigned integer values.
```
Comment (kainino0x) [inline spec/index.bs:7980]: needs to define rounding of "halfway" values - see WebIDL rules here:
https://webidl.spec.whatwg.org/#abstract-opdef-converttoint

Also need to decide whether we clamp values that are out of range (probably?) or do something else with them

File: spec/index.bs [hunk 5]
```diff
@@ -7960,6 +7960,25 @@ dictionary GPURenderPassColorAttachment {
...
+        {{GPURenderPassColorAttachment/view}}.
+
+        Let |colorAttachmentFormat| be
+        {{GPURenderPassColorAttachment/view}}.{{GPUTextureView/[[descriptor]]}}.{{GPUTextureViewDescriptor/format}}.
+        - If the components of |colorAttachmentFormat| are floating point values, the members of
+            {{GPURenderPassColorAttachment/clearValue}} will be converted into floating point
+            values.
+        - If the components of |colorAttachmentFormat| are signed or unsigned normalized values,
+            the members of {{GPURenderPassColorAttachment/clearValue}} will first be clamped
+            into the range [0, 1], then converted into floating point values, and finally
```
Comment (kainino0x) [inline spec/index.bs:7975]: intermediate conversion to "floating point value" not necessary because it's already a floating-point value

File: spec/index.bs [hunk 6]
```diff
@@ -7960,6 +7960,25 @@ dictionary GPURenderPassColorAttachment {
...
         if {{GPURenderPassColorAttachment/loadOp}} is not {{GPULoadOp/"clear"}}.
 
+        The members of {{GPURenderPassColorAttachment/clearValue}} are all double values, so
+        they will first be converted to the fully qualified format type of
+        {{GPURenderPassColorAttachment/view}} before being set as the clear value of
+        {{GPURenderPassColorAttachment/view}}.
+
+        Let |colorAttachmentFormat| be
+        {{GPURenderPassColorAttachment/view}}.{{GPUTextureView/[[descriptor]]}}.{{GPUTextureViewDescriptor/format}}.
+        - If the components of |colorAttachmentFormat| are floating point values, the members of
```
Comment (kainino0x) [inline spec/index.bs:7970]: > If the components

nit: Since each component can be a slightly different type, e.g. rgb10a2unorm, I would tweak the language here slightly to define in terms of a single component (scalar input value and component type) and say it applies to each component separately.

File: spec/index.bs (hunk 5 above)
Comment (Jiawei-Shao) [reply spec/index.bs]: I'd like to convert the double values to float first because on D3D12 the clear value can only be float, and I think we can align the behaviors on all the backends if we always use float.

File: spec/index.bs (hunk 6 above)
Comment (Jiawei-Shao) [reply spec/index.bs]: Now I am using the concept of "channel" (defined in WGSL SPEC) here.

File: spec/index.bs (hunk 2 above)
Comment (Jiawei-Shao) [reply spec/index.bs]: I think we can always convert the clear values into the ones that can be accepted by all the backends before applying these clear values in the native APIs. For floating point formats, we can always convert the clear values into 32-bit-float ones, which are accepted by all the native APIs (especially on D3D12).

File: spec/index.bs (hunk 5 above)
Comment (kainino0x) [reply spec/index.bs]: ah, in that case we would need to specify the intermediate format is 32-bit/single-precision float (as you now have). But I don't think we should spec the extra conversion as it could result in double-rounding. Instead I would specify the conversion directly from float64 to the format type, and make D3D12 implementations responsible for picking the float32 value that results in the proper end result (or [polyfilling](https://github.com/gpuweb/gpuweb/issues/1085#issuecomment-703858922) when necessary).

File: spec/index.bs (hunk 2 above)
Comment (kainino0x) [reply spec/index.bs]: For float formats we still have to define how to get all the way from a clear-value to what bits get put in the texture. So for example we need to fully define a conversion to float16 (which webidl doesn't have, but we can use the same rules).

File: spec/index.bs (hunk 5 above)
Comment (Jiawei-Shao) [reply spec/index.bs]: I've updated my latest PR with your suggestions here. PTAL, thanks!

File: spec/index.bs [hunk 7]
```diff
@@ -7960,6 +7960,28 @@ dictionary GPURenderPassColorAttachment {
...
+        Let |colorAttachmentFormat| be
+        {{GPURenderPassColorAttachment/view}}.{{GPUTextureView/[[descriptor]]}}.{{GPUTextureViewDescriptor/format}}.
+        A |colorAttachmentFormat| has up to four |channel|s: `r`, `g`, `b`, and `a` (each |channel|
+        contains one scalar), normally corresponding to the concepts of red, green, blue, and alpha
+        channels of {{GPURenderPassColorAttachment/clearValue}}.
+
+        For each |channel| of |colorAttachmentFormat|:
+        - If the |channel| contains a {{float}}, signed or unsigned normalized value, the
+            corresponding member of {{GPURenderPassColorAttachment/clearValue}} will be converted
+            into the format type of |channel|.
```
Comment (kainino0x) [inline spec/index.bs:7977]: This is still underspecified:
- Still need to specify that normalized values are clamped and how "halfway" values are rounded.
- Conversion to `{{float}}` (IDL float - actually should be `{{unrestricted float}}`) isn't enough because that's just float32, but we have float16, ufloat10, and ufloat11 as well. It's OK if, for now, we just say something like "Issue: Define our own conversions for float16/ufloat10/ufloat11 following the [same pattern](https://webidl.spec.whatwg.org/#es-unrestricted-float) as {{unrestricted float}}".

Comment (Jiawei-Shao) [reply spec/index.bs]: > Still need to specify that normalized values are clamped and how "halfway" values are rounded.

Oh in [D3D SPEC](https://microsoft.github.io/DirectX-Specs/d3d/archive/D3D11_3_FunctionalSpec.htm#FLOATtoUNORM) there is also no definition of "how 'halfway' values are rounded", and I remember it is a known issue that float-to-unorm-conversion is not always precise (e.g. 0.5 may be converted to either 127 or 128).

In addition, specifying a float32 value as the clear value of float/snorm/unorm formats is well supported on all the backends, and here I think we can just focus on how to handle int/uint formats.

**Comment by kainino0x** [comment]
editors meeting agenda: choose between strict conversions (WebIDL style) and loose conversions (shader style) and propose to `tacit resolution queue`

File: spec/index.bs (hunk 7 above)
Comment (kainino0x) [reply spec/index.bs]: We don't need D3D to define how halfway values are rounded. We could spec it tightly and implementations can make it work:
- receive js double 0.5
- webgpu implementation knows the format is unorm8 and deterministically chooses a value of 128 (for example)
- webgpu implementation computes `128f / 255f` and passes that as the D3D clear value

That said, now that I'm thinking about it, this would be a lot stricter than the conversion when a shader writes an f32 to a texture. So there's not much reason to define it very strictly, as much as I want to. (We can always tighten it later.)

It also occurs to me that well-defining conversions is easier said than done for a few cases. In particular the `double` -> `f32` conversion defined by WebIDL is stricter than [the C definition](https://stackoverflow.com/q/65610811) of double->float cast, and if we define all our conversions in the WebIDL style they'll be stricter than shader f32->texture writes.

So now I'm thinking I want to propose we define the conversions here in the shader style. Unfortunately I don't know what that means exactly, but it at least means that 0.5 -> 127 and 128 are both valid. Not sure about f64->f32/f16/uf10/uf11. (Is truncation valid, as it seems to be in C? Or do you have to round to the closest value when possible?)

Comment (kainino0x) [reply spec/index.bs]: Editors agreed looser conversions are fine, up to whatever we allow when converting shader f32 into a render attachment... which we don't specify yet\*. But we can define this in terms of that.

Note that conversions to integer formats are still fully strict: AFAICT in both C and WebIDL they always truncate, so there's no ambiguity about what integer you get from a double. We only have to define the behavior (clamp) when the double is out of range of the target int type.

\* We know it's not super strict (e.g. 0.5 can be either 127 or 128), but it's also probably not quite as loose as C: at least in D3D11, there's this stipulation on "Data Invertability":
https://microsoft.github.io/DirectX-Specs/d3d/archive/D3D11_3_FunctionalSpec.htm#19.1.2%20Data%20Invertability
which says that going from memory -> shader -> memory won't change the value (except for snorm `-1.0`)

Comment (kainino0x) [reply spec/index.bs]: > Note that conversions to integer formats are still fully strict: AFAICT in both C and WebIDL they always truncate, so there's no ambiguity about what integer you get from a double.

Oh yeah, IDL defines clamping-casts to int as rounding, not truncating...

Comment (kainino0x) [reply spec/index.bs]: I'll propose making integers use a _non_ clamping cast, like conversion to IDL `long` (rather than `[EnforceRange] long` or `[Clamp] long`.

Comment (kainino0x) [reply spec/index.bs]: Revised accordingly.

**Comment by kainino0x** [comment]
for tacit resolution queue(?):
The conversion defined here are NOT the ones we originally decided on when we discussed this in #965. Instead, it defines them in terms of the "Output Merging" step: that is, the result is the same as if you wrote the same value from WGSL. (First converting to f32/i32/u32 using WebIDL's rules.)

This means we won't have to define two sets of conversion rules (one for clear values, one for output merging).

**Comment by kainino0x** [comment]
@Jiawei-Shao could you also review the definitions I proposed in my revision?

File: spec/index.bs [hunk 8]
```diff
@@ -7967,6 +7967,45 @@ dictionary GPURenderPassColorAttachment {
...
+            value |value| in {{GPURenderPassColorAttachment/clearValue}}:
+
+            1. If |componentType| is a:
+
+                <dl class=switch>
+                    : floating-point type or normalized type
+                    :: Convert |value| [=converted to an IDL value|to an IDL value=] of type {{unrestricted float}} (`f32`).
+                    : signed integer type
+                    :: Convert |value| [=converted to an IDL value|to an IDL value=] of type {{long long}} (`i32`).
+                    : signed integer type
```
Comment (Jiawei-Shao) [inline spec/index.bs:7992]: This should be "unsigned integer type".

**Comment by Jiawei-Shao** [comment]
@kainino0x LGTM!

I just found I couldn't approve my own PR.

File: spec/index.bs (hunk 8 above)
Comment (kainino0x) [reply spec/index.bs:7992]: Nice catch, thanks!

Comment (kainino0x) [reply spec/index.bs:7992]:
```suggestion
                    : unsigned integer type
```
